### PR TITLE
S3 offline logic

### PR DESCRIPTION
**Pull request by pbywater**
Relates to #163 #170 
- Saves base64 encoded url to localStorage whilst offline
- Decodes and uploads to database and s3 once back online
- Limits users to one offline media upload and alerts them if they are trying to do more
- Refactors s3 code so it works with audio too (online and offline)

### CODE REVIEW

**Review by ghost — APPROVED** [review]

File: public/scripts/s3-helpers.js [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  if (file) {
+    reader.readAsDataURL(file);
+  }
+}
+
+function retrieveFile(fileSource) {
+  const file = localStorage.getItem(`toSaveFrom${fileSource}`);
+  const fileName = localStorage.getItem(`${fileSource}fileName`);
+  if (file && fileName) {
+    const retrievedFile = dataURLtoFile(image, fileName);
```
Comment (des-des) [inline public/scripts/s3-helpers.js]: image is undefined?

File: public/scripts/s3-helpers.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+  return null;
+}
+
+function addTagAndHeadingToDB(tag, heading, imageId) {
+  $.ajax({
+    method: 'PUT',
+    url: 'memory-input-photo',
+    data: { tag, heading, imageId },
+    success: () => { uploadPending = false; },
```
Comment (des-des) [inline public/scripts/s3-helpers.js:29]: No global scope please. Where has this come from?

Not sure what is going on / how I missed this, but you want to be writing your code with require / module.exports.

Everything in the top level of this file is declared in global scope. Further, this code is not being passed though babel (you browserify script does this).

That means your arrow functions are not being converted into es5 and will break some browsers

File: public/scripts/s3-helpers.js [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  });
+}
+
+function dataURLtoFile(dataurl, filename) {
```
Comment (des-des) [inline public/scripts/s3-helpers.js]: please link to the source of code that you have copy pasted

File: public/scripts/s3-uploads.js [hunk 4]
```diff
@@ -1,61 +1,91 @@
...
-  const xhr = new XMLHttpRequest();
-  xhr.onreadystatechange = () => {
-    if (xhr.readyState === 4) {
-      if (xhr.status === 200) {
-        const response = JSON.parse(xhr.responseText);
-        uploadFile(file, response.signedRequest, response.url);
-        imageUploadPending = true;
-        updateTagAndHeading(response.imageId);
+function s3handling(fileSource) {
+  (function () {
```
Comment (des-des) [inline public/scripts/s3-uploads.js]: This iife is nonsense.

You need to use require / module.exports. This is no no good we are going backwards!

File: public/scripts/s3-uploads.js [hunk 5]
```diff
@@ -1,61 +1,91 @@
...
+  let uploadPending = false;
+
+  function getSignedRequest(file) {
+    const xhr = new XMLHttpRequest();
+    xhr.onreadystatechange = () => {
+      if (xhr.readyState === 4) {
+        if (xhr.status === 200) {
+          const response = JSON.parse(xhr.responseText);
+          uploadFile(file, response.signedRequest, response.url);
+          uploadPending = true;
```
Comment (des-des) [inline public/scripts/s3-uploads.js:29]: :scream:

**Review by des-des — CHANGES_REQUESTED** [review]
@pbywater 

Hi I know are kinda tight on time etc etc, but I would **strongly** recommend you refactor this to stop using global vars. Its evil, fragile and unmaintainable.

Further, the es6 code you have written will crash older browsers if you are not passing it through your build process.

Sorry I did not pick this up earlier..

File: public/scripts/s3-uploads.js (hunk 5 above)
Comment (pbywater) [reply public/scripts/s3-uploads.js:29]: Haha - this was all the code that was merged before!! I will take a look and refactor

File: public/scripts/s3-helpers.js [hunk 6]
```diff
@@ -0,0 +1,58 @@
+function saveFileToLocalStorage(file, fileSource) {
```
Comment (des-des) [inline public/scripts/s3-helpers.js:1]: cool, this file should live in src

File: src/client/main.js [hunk 7]
```diff
@@ -1,4 +1,8 @@
 require('./d3/d3.js');
+require('./../../public/scripts/main.js');
+require('./../../public/scripts/s3-helpers.js');
+require('./../../public/scripts/s3-uploads.js');
+require('./../../public/scripts/sign-up.js');
```
Comment (des-des) [inline src/client/main.js:5]: yeah, all these files should now move to src!

File: public/service-worker.js [hunk 8]
```diff
@@ -25,7 +25,9 @@ const filesToCache = [
   '/index.html',
   '/main.css',
   '/bundle.js',
-  '/scripts/main.js'];
+  '/scripts/main.js',
```
Comment (des-des) [inline public/service-worker.js]: There should only be 1 bundle? No other scripts?

File: public/scripts/s3-uploads.js [hunk 9]
```diff
@@ -1,61 +1,90 @@
-(function () {
-  document.getElementById('camera-input').onchange = function () {
-    const files = document.getElementById('camera-input').files;
+const { saveFileToLocalStorage, addTagAndHeadingToDB, removeItemsFromStorage } = require('./s3-helpers.js');
+
+function s3handling(fileSource) {
```
Comment (des-des) [inline public/scripts/s3-uploads.js:3]: can we export this, then call it in main. (if that makes sense, right now this function gets executed when it is required, but that can be very hard to follow)

**Review by des-des — CHANGES_REQUESTED** [review]
